core/src/parking_lot.rs

```diff
@@ -81,36 +92,75 @@ impl Clone for Bucket {
     }
 }
 
-struct FairTimeout {
-    // Next time at which point be_fair should be set
-    timeout: Instant,
+#[cfg(not(all(target_env = "sgx", target_vendor = "fortanix")))]
+mod fair_timeout {
+    use super::*;
+    use core::time::Duration;
+    #[cfg(not(feature = "i-am-libstd"))]
+    use rand::{rngs::SmallRng, FromEntropy, Rng};
 
-    // Random number generator for calculating the next timeout
-    rng: SmallRng,
-}
+    pub struct FairTimeout {
+        // Next time at which point be_fair should be set
+        timeout: Instant,
 
-impl FairTimeout {
-    #[inline]
-    fn new() -> FairTimeout {
-        FairTimeout {
-            timeout: Instant::now(),
-            rng: SmallRng::from_entropy(),
+        // Random number generator for calculating the next timeout
+        #[cfg(not(feature = "i-am-libstd"))]
+        rng: SmallRng,
+    }
+
+    impl FairTimeout {
+        #[inline]
+        pub fn new() -> FairTimeout {
+            FairTimeout {
+                timeout: Instant::now(),
+                #[cfg(not(feature = "i-am-libstd"))]
+                rng: SmallRng::from_entropy(),
+            }
+        }
+
+        // Determine whether we should force a fair unlock, and update the timeout
+        #[inline]
+        pub fn should_timeout(&mut self) -> bool {
+            let now = Instant::now();
+            if now > self.timeout {
+                #[cfg(not(feature = "i-am-libstd"))]
+                let ns_offset = self.rng.gen_range(0, 1000000);
+                #[cfg(feature = "i-am-libstd")]
+                let ns_offset = 500000;
+                self.timeout = now + Duration::new(0, ns_offset);
+                true
+            } else {
+                false
+            }
         }
     }
+}
 
-    // Determine whether we should force a fair unlock, and update the timeout
-    #[inline]
-    fn should_timeout(&mut self) -> bool {
-        let now = Instant::now();
-        if now > self.timeout {
-            self.timeout = now + Duration::new(0, self.rng.gen_range(0, 1000000));
-            true
-        } else {
-            false
+#[cfg(all(target_env = "sgx", target_vendor = "fortanix"))]
+mod fair_timeout {
+    pub struct FairTimeout {
+        counter: usize,
+    }
+
+    impl FairTimeout {
+        #[inline]
+        pub fn new() -> FairTimeout {
+            FairTimeout { counter: 0 }
+        }
+
+        // Determine whether we should force a fair unlock, and update the timeout
+        #[inline]
+        pub fn should_timeout(&mut self) -> bool {
+            // On SGX it's expensive to call `Instant::now()`. So we resort
+            // to make every 100 call be true.
+            self.counter += 1;
+            self.counter % 100 == 0
```

Comment:
I just pulled this number out of thin air. Maybe we want the interval to be way different, or maybe it's sufficient to just always return `false`. But I had to start somewhere.

Reply:
That feels very suspicious and should probably be reverted. Are you sure that `Instant::now` is a performance bottleneck. Keep in mind that we are only calling it on the slow path where we are performing a syscall to unpark a thread anyways.

Reply:
And every time we resize the hashtable. Which admittedly should not be very often after a few initial resizes. But we don't call it once per resize, but once per bucket per resize. However, those initial calls can be reduced to one per resize without actually changing this algorithm, that is true!

According to @jethrogb `Instant::now()` takes ~6us on SGX. As compared to ~100ns on Linux. On their machine. But I did not yet get any feedback as to whether or not this conditional compilation actually affected any benchmark. However, from debugging/tracing their runs of the test suite they saw the library being stuck within `Instant::now()` quite frequently. Mind, that was before the duplication bug for `HASHTABLE` was fixed, so maybe things are different now.

I have not yet had time to reinstall my lab machine so I can get SGX running and actually benchmark this myself.

Reply:
SGX usercalls are very slow and should be avoided if possible. If you turn every `send` call into two usercalls, you've now made waking up twice as slow (12µs instead of 6µs). However, this issue seems relatively minor compared to the next point.

SGX is a technology for running software in untrusted environments, by running that software in a so-called secure enclave. Any outside input (such as time, network data, etc.) must be assumed to be under the control of an adversary. By making scheduling decisions based on the return value of `insecure_time`, the attacker is getting fine-grained control over which tasks can make forward progress and which tasks can be held back. This is not acceptable. All scheduling decisions must be made by the code running in the enclave. The attacker may still interfere with scheduling, but it can't influence the ordering of synchronized events in the enclave.

What does this mean for `parking_lot`? I think `FairTimeout::should_timeout` should always return `false`. But more generally, it means I clearly don't understand the scheduling mechanisms in this crate enough to determine if it's secure. I'm not sure if I can approve swapping out the `std::sync` implementation for the SGX platform with this crate, without doing a more thorough review of the entire codebase.

Reply:
It's not doing it for every wake up. Only on the slow path. `parking_lot` has a fast and a slow path and a fast path. It always tries to use the fast path, but if that fails, it has to fall back to full OS parking etc.

The value returned from `should_timeout` does not dictate very much about scheduling in general at all. It just makes sure there is a fair lock handover approximately every 500µs. I think the worst thing the attacker could do is either disable the fairness (which your suggestion for a hardcoded `false` would also do) or slightly control when it's triggered. But it has no control at all over who will get the lock when it uses fairness vs not using fairness etc.

Reply:
> the attacker is getting fine-grained control over which tasks can make forward progress and which tasks can be held back

No they don't get this fine-grained control. As I described above, the worst thing they could do is control which of the unlock mechanisms are being used: fair or not fair. But they would not have any influence over what thread is next in line if they forced fairness, so they could not control who is going to take over the lock and continue executing. And in the case of a non fair unpark it's just a race between the unparked thread and any other thread that might be interested in the lock. The time does not affect the race in any way that I can think of.